### ISSUE TITLE
Expose evaluate() for user-provided params

### ISSUE THREAD

**Issue by dae** [ankitects/anki#4062]
1) Evaluate button is removed, but the underlying routine is still exposed in our Python API. Providing a bit of example code that users can copy+paste to evaluate their tweaks would be a nice gesture for affected users.

_Originally posted by @dae in https://github.com/ankitects/anki/issues/3962#issuecomment-2927298959_

- Currently, it appears we're only exposing the new implementation, and not the old one.
- Some sample code for users would be good

**Comment by JSchoreels** [comment]
I don't even understand why it has been removed ? 

It wasa nice way to see if the improvement was big enough to take the hit of a re-optimize backlog.

Now I might get hundreds of review to do without being even able to know what's the pros ?

**Comment by markonm** [comment]
Maybe I was mistaken, but I was using the old evaluate button as a reference when deciding if I needed to create new presets for different topics. The new "health check" always displays the same message and takes significantly more time for calculation while providing nothing of relevance to me.

I hope the information from the evaluate is eventually put into healthy check window or indicated somewhere else without needing to use add-ons or debug console.

**Comment by Expertium** [comment]
The goal of this change is to make Anki easier to understand and less confusing for newcomers. Displaying a bunch of abstract numbers does the opposite of that. Plus, with the way health check works, it can display a warning for one person and not display a warning for another person even if their log loss and RMSE are the same, because it uses a correction based on n(reviews) and true retention. So if anything, bringing the log loss and RMSE numbers back will be even more confusing because the exact same values may result in a warning for one user but not the other user.

The old Evaluate will only be available via the debug console. If you have a better idea - feel free to share it, but I don't think there is a way to both hide the old Evaluate from new users and make it easy to access for power users at the same time.

Well, technically, there is, but Dae hates it: split UI. Aka Beginner/Advanced switch that changes UI and hides some settings and toggles in the Beginner mode. Then we could only show the old Evaluate in the Advanced mode. And it would also solve, like, 20-30 different UI problems.

**Comment by JSchoreels** [comment]
If we focus on UX, then explain now how users :
- Now if they gain anything by splitting decks into different subsets or if they could just merge two into one?
- If an optimized set of parameters is worth the backlog to do in the next days?

I'm sorry to storm about this @dae, but I find unfortunate that those past months are filled with so many different debates leading to people disengaging from discussion until we realize some key component **actual users** (and not the ones some in their mind try to romantize) use daily.

At this rate, I'm even worry about FSRS becoming the default option and optimization even becoming automatic. It just feel like some people (well, mostly one)  keep on repeating "for the uers !" while doing everything possible to prove he's smarter than all the user base, actively forcing his way upon us all.

Focus on your users, not on the one you imagine you'll have.

**Comment by Expertium** [comment]
@JSchoreels 

We don't have a good way of determining which decks should use separate presets, and I don't think that using the old Evaluate is a good idea. Log loss is correlated with retention, RMSE is correlated with retention and with n(reviews). So lower/higher log-loss or RMSE alone doesn't indicate that splitting/merging is better. Simplest example: merging two decks can result in a lower RMSE than the RMSE of either deck just because n(reviews) has increased. I don't know how to solve this problem, it would require some new metrics that aren't correlated with either retention or n(reviews), and we don't have such metrics. I'm not sure if it's even theoretically possible to make such metrics. In the health check, I just use an *ad hoc* correction.
Maybe in the future me and Jarrett will think of some way to group cards based on how similar their review histories are or something, but that won't be today, or tomorrow.

As for the second part of your comment, I think you severely overestimate how much complexity an average user can handle before they just give up on Anki, or at least on changing any settings in Anki.

---

Btw, Dae, with the split Beginner/Advanced UI we would never have to argue about whether Anki should be oriented towards average, non-tech-savvy users or power users: it could be oriented towards both! Right now the problem is that what power users want is not what average users want, but with a Beginner/Advanced switch, each user can just stick to whatever he is more comfortable with - a simplified UI with the most basic functionality or a more complex UI with more settings and options. With a split UI, removing things from the Beginner UI mode won't make life worse for power users, and adding more features into the Advanced UI mode won't make Anki more confusing for average users.

Beginner/Advanced switch is as close to "holy grail" as it gets, unless we get into the sci-fi territory of things like "Personal AI assistant that can explain every button and every setting to you in a way that you will understand, and can adjust anything in Anki to your liking, like a genie". Having two UI modes doesn't require sci-fi technology and forever eliminates the friction between average users and power users.

**Comment by Danika-Dakika** [comment]
I'm still disappointed that the end result was that the Evaluate button being removed. But It also doesn't sound like this other method of accessing it will be usable for support. Is the idea that the Health Check will be self-explanatory and eliminate the need for posting questions/support intervention?

**Comment by user1823** [comment]
> Is the idea that the Health Check will be self-explanatory and eliminate the need for posting questions/support intervention?

@Danika-Dakika, I don't think it will eliminate the need for posting questions. But, is the Health Check worse in any way (from a support perspective) than the old Evaluate?

If a user complains about something, they can be asked to enable the health check (assuming they already disabled it) and click Optimize. This ensures that they have the most optimal parameters. In addition, they are informed if FSRS is not able to adjust to their review history, which is what a supporter was earlier supposed to do when the users posted their RMSE/logloss values.

Am I missing something?

<hr>

> Know if they gain anything by splitting decks into different subsets or if they could just merge two into one?

@JSchoreels @markonm 

To know this, you would need to split the deck into two, optimize parameters on each subset and then evaluate the generated parameters on the respective subsets.

This means that the parameters are optimized and evaluated on the same data. So, even if the metrics of the subset are better, you won't be able to know if this is because of overfitting or because the parameters are actually better.

We need those parameters which can generalize to unseen reviews (reviews that are going to happen in the future). We don't want those parameters that perform extremely well on the current data because of overfitting but perform poorly on subsequent reviews.

So, in my opinion, the old Evaluate shouldn't be used to decide whether to split presets because we can't know if splitting actually generates better parameters.

Any counter arguments are welcome.

<hr>

> If an optimized set of parameters is worth the backlog to do in the next days?

@JSchoreels, this is an interesting argument. I never thought of it before. But, I think that we have a better alternative.

Let's say that the new parameters are significantly better than the previous ones. But, does this justify doing a card that was due tomorrow and had an ivl of 56 days, which got an interval of 55 days after rescheduling and got due today? IMO, it doesn't because even though the metrics on the whole collection improved drastically, 55d vs 56d is not that significant.

So, what's the alternative?
Instead of deciding whether to reschedule or not based on the metrics of the whole preset, decide based on the change in the stability of each card.

How to do that?
In the config of the FSRS helper add-on, set `reschedule_threshold` to something like 0.3 (adjust based on your preferences) and then use the add-on's reschedule option instead of the built-in one.

As I said before, counter arguments are welcome.

**Comment by jucor** [comment]
FWIW, as a nerd Anki user and a machine learning researcher, I *loved* being able to see the loss and RMSE values and check how much they had changed after optimization. That gave me more confidence in using FSRS, and in the evaluation of the loss function (whether my latest reviews made a big change to it or not).
Plus it gave a nice "Yay FSRS!" when I updated it and saw them go down :)
Having now just a health check that essentially says "Ruh roh, can't fit to your memory, sorry" makes me quite nervous: shall I _not_ be using FSRS in this case? 
I'm absolutely fine with the usual  "oooh, that's intricate and very thought through, I'm going to need to do a lot of reading to understand this!" nature of Anki, which comes with the territory. 
But today is the first time Anki causes me worries of the "errr, that's too opaque, and more dumbed-down than I would want, what's going on" sort.
